### PR TITLE
Add basic support for non Gregorian calendars

### LINKED ISSUE #1966: Alternative calendar support in time block
There are many calendars other than the Gregorian calendar that are used in the world. For example, the ecmascript and web platform support [19 of them](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/getCalendars#supported_calendar_types). Would you accept a PR that add support for these calendars in the time block (or a new block) using the [icu4x](https://github.com/unicode-org/icu4x) crate?

### PR DESCRIPTION
**Pull request by HKalbasi**
fix #1966

### CODE REVIEW

**Comment by HKalbasi** [comment]
CI failure is due `locid` which is a module in the `icu` crate. What should I do about it?

**Comment by MaxVerevkin** [comment]
> CI failure is due `locid` which is a module in the `icu` crate. What should I do about it?

You can edit `cspell.yaml` to add this new word.

**Comment by HKalbasi** [comment]
I replaced the `icu` crate with `icu_calendar` and `icu_datetime`. Now 18 dependencies are added, 14 of them are icu internal crates.

File: src/blocks/time.rs [hunk 1]
```diff
@@ -31,6 +31,22 @@
 //!
 //! # Icons Used
```
Comment (bim9262) [inline src/blocks/time.rs:32]: I'd keep this after both examples

File: src/formatting/formatter.rs [hunk 2]
```diff
@@ -535,25 +561,40 @@ impl DatetimeFormatter {
...
+                    },
+                }
+                .to_string(),
+                DatetimeFormatter::Icu { locale, length } => {
+                    let date = icu_calendar::Date::try_new_iso_date(
+                        datetime.year_ce().1 as i32,
+                        datetime.month() as u8,
+                        datetime.day() as u8,
+                    )
+                    .expect("Current date should be a valid date");
```
Comment (bim9262) [inline src/formatting/formatter.rs:588]: I think we should use `.error` instead of `.expect`.

**Comment by HKalbasi** [comment]
I found out that I didn't handle the timezone (by using the patch on my machine in a full day), and fixed that in the latest commit.

**Comment by MaxVerevkin** [comment]
Binary size change: stripped 17M -> 31M, unstripped 24M -> 39M. That's too much... Would you be okay to put this feature behind a feature flag?

**Comment by HKalbasi** [comment]
I made it optional with a feature flag and I'm happy with that. Instead of / in addition to that, we can split the [locale data from the binary](https://github.com/unicode-org/icu4x/blob/main/docs/tutorials/data_management.md), though it will complicate things.

**Review by MaxVerevkin — APPROVED** [review]

**Review by bim9262 — APPROVED** [review]
@HKalbasi I have used the approach you mentioned to split the locale data, and while it's doable it's a lot of work to slim down the binary size.

**Pull request merged**